packages/schematics/src/utils/module-utils.ts

```diff
@@ -0,0 +1,218 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {
+  findNodes,
+  insertAfterLastOccurrence
+} from '@schematics/angular/utility/ast-utils';
+import {
+  Change,
+  InsertChange,
+  NoopChange,
+  ReplaceChange,
+  RemoveChange
+} from '@schematics/angular/utility/change';
+
+import * as ts from 'typescript';
+import { offset, findClass } from './ast-utils';
+
+export function addClass(
+  source: ts.SourceFile,
+  modulePath: string,
+  clazzName: string,
+  clazzSrc: string
+): Change {
+  if (!findClass(source, clazzName, true)) {
+    const nodes = findNodes(source, ts.SyntaxKind.ClassDeclaration);
+    return insertAfterLastOccurrence(
+      nodes,
+      offset(clazzSrc, 0, true),
+      modulePath,
+      0,
+      ts.SyntaxKind.ClassDeclaration
+    );
+  }
+  return new NoopChange();
+}
+
+export function addUnionTypes(
+  source: ts.SourceFile,
+  modulePath: string,
+  typeName: string,
+  typeValues: string[]
+) {
+  const target: ts.TypeAliasDeclaration = findNodesOfType(
+    source,
+    ts.SyntaxKind.TypeAliasDeclaration,
+    name => name === typeName
+  );
+  if (!target) {
+    throw new Error(`Cannot find union type '${typeName}'`);
+  }
+
+  const node = target.type as ts.TypeReferenceNode;
+
+  // Append new types to create a union type...
+  return new InsertChange(
+    modulePath,
+    node.end,
+    ['', ...typeValues].join(' | ')
+  );
+}
+
+/**
+ * Add 1..n enumerators using name + (optional) value pairs
+ */
+export function addEnumeratorValues(
+  source: ts.SourceFile,
+  modulePath: string,
+  enumName: string,
+  pairs: NameValue[] = []
+): Change[] {
+  const target = findNodesOfType(
+    source,
+    ts.SyntaxKind.EnumDeclaration,
+    name => name === enumName
+  );
+  const list = target ? target.members : undefined;
+
+  if (!target) {
+    throw new Error(`Cannot find enum '${enumName}'`);
+  }
+
+  return pairs.reduce((buffer, it) => {
+    const addComma = !(list.hasTrailingComma || list.length === 0);
+    const member = it.value ? `${it.name} = '${it.value}'` : it.name;
+    const memberExists = () => {
```

Comment:
why is this a function? it looks like it's called once without any arguments.

Reply:
reduction of name/value pairs to array of InsertChange changes.